content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+
+```python
+def dom_tree_copy_elimination(func: dict):
+    """Eliminate copies using dominator-tree-based propagation.
+
+    Walks the dominator tree carrying a rename environment. For each `id`
+    instruction, we establish a rename from dest to source and drop the
+    instruction.
+    """
+    def walk(block_name: str, rename: Dict[str, str]):
+        block = bmap[block_name]
+        new_instrs = []
+
+        for instr in block.get("instrs", []):
+            op = instr.get("op")
+
+            # Rewrite arguments using rename map (but NOT phi args)
+            if "args" in instr and op != "phi":
+                instr["args"] = [canon(a, rename) for a in instr["args"]]
+
+            # Handle id instructions
+            if op == "id":
+                dest = instr.get("dest")
+                if dest not in phi_dests:
+                    src = canon(args[0], rename)
+                    rename[dest] = src
+                    continue  # Drop the id instruction
+
+            new_instrs.append(instr)
+
+        block["instrs"] = new_instrs
+
+        # Recurse with copied environment
+        for child in dom_tree.get(block_name, []):
+            walk(child, rename.copy())
+```
+
+#### 2.4.2 Why This is Safe
+
+The key insight is that renames established in a dominator are valid in all dominated blocks. By copying the environment before recursing into children, we ensure that renames from one branch don't leak into sibling branches.
+
+**Example:**
+```
+     B0: x = 5
+    /  \
+   B1   B2
+   y=x  z=x
+```
+
+If we're at B0 and establish `x → 5`, both B1 and B2 see this rename. But renames in B1 don't affect B2 because they receive separate copies of the environment.
+
+#### 2.4.3 Preserving Phi Destinations
+
+We explicitly preserve phi destinations because they are **merge points**. Even if a phi produces `x = phi(y, z)` where `y` and `z` have the same value, we can't eliminate the phi — it defines a new SSA name that may be used later.
+
+### 2.5 Fixed Parallel Copy Sequentialization {#25-fixed-parallel-copy-sequentialization}
+
+#### 2.5.1 The Problem
+
+When lowering phi nodes to copies, we face the **parallel copy problem**. Phi nodes represent parallel assignments:
+
+```
+# At block B with predecessors P1, P2:
+x = phi(a from P1, b from P2)
+y = phi(c from P1, d from P2)
+```
+
+These become parallel copies at the end of each predecessor:
+```
+# End of P1:
+(x, y) := (a, c)  # Parallel!
+```
+
+If we naively sequentialize:
+```
+x = a
+y = c
+```
+
+This is fine. But consider:
+```
+(x, y) := (y, x)  # Swap!
+```
+
+Naive sequentialization fails:
+```
+x = y   # x gets y's value
+y = x   # y gets... x's new value (wrong!)
+```
+
+#### 2.5.2 The Algorithm
+
+The fixed algorithm handles dependencies correctly:
```

Comment:
You have described a potential problem, but you haven't described _what actually has this problem_. What are we fixing here?